internal/types/document_validation.go

```diff
@@ -79,5 +79,14 @@ func (d *Document) ValidateData() error {
 		return newValidationError(fmt.Errorf("invalid document: document must contain '_id' field"))
 	}
 
+	v := must.NotFail(d.Get("_id"))
+
+	switch v.(type) {
+	case Array, *Array:
+		return fmt.Errorf("The '_id' value cannot be of type array")
```

Comment:
Ok, after reading the whole PR, I got what you asked yesterday :) 

I suggest adding the `code common.ErrorCode` field to `ValidationError` struct, so we can either set it as `ErrFailedToParse` or that new const for `53` for those new cases. 

Then we can still use `ValidationError` for all the cases here, and we can extend it if we need more error codes in the future.

@noisersup @chilagrow @w84thesun @AlekSi, what do you think?

Reply:
so the ValidationErrors's `Error()` function will return different output when the code is equal to `ErrFailedToParse` and different if it's set to `53`?

Reply:
```
if errors.As(err, &valErr) {
	return common.NewErrorMsg(valErr.Code(), err.Error())
}
```

where `valErr.Code()` calls a new method `Code` that returns the `code` field from the structure.

Reply:
And `Error` func can be rewritten, yes (e.g. we can remove `Invalid document, reason` part).

Reply:
No need to handle different error codes differently, let's rather make`ValidateData` more universal (and remove everything that doesn't let it be universal).

Reply:
Yay!

Reply:
@noisersup, and yesterday I didn't get what you meant, sorry for that.

Reply:
No worries!

Reply:
Using `common.ErrorCode` here creates import cycle. I suggest moving ValidationError into `common` package

Reply:
Let's do the simplest thing now. We could always replace it with method/interface/types/etc later.

There:
```go
type ValidationErrorCode int

var (
  ErrBadID ValidationErrorCode = iota
)

type ValidationError struct {
  Code ValidationErrorCode
  reason error
}
```

Somewhere in handler layer:

```go
func checkError(err error) error {
switch err := err.(type) {
  case *ValidationError:
    switch err.Code {
      case types.ErrBadID:
        return ErrInvalidID
    }
}
```